src/java/main/org/apache/zookeeper/server/NIOServerCnxn.java

```diff
@@ -479,7 +479,7 @@ private boolean checkFourLetterWord(final SelectionKey k, final int len)
         // We take advantage of the limited size of the length to look
         // for cmds. They are all 4-bytes which fits inside of an int
         String cmd = FourLetterCommands.getCmdMapView().get(len);
-        if (cmd == null) {
+        if (cmd == null || !FourLetterCommands.getWhiteListedCmdView().contains(cmd)) {
```

Comment:
I think the original comment was not clear but I think it is a good catch - instead of return false here we return true because the semantic of checkFourLetterWord is we only return false if 4lw is not found, and in that case the caller will think this is a client message and proceed allocate buffer etc work (iiuc that was what the "it should be processed in that way only" meant.).